### ISSUE TITLE
`insert_final_newline == false` --- strip EOL, or leave untouched?

### ISSUE THREAD

**Issue by cxw42** [editorconfig/editorconfig#475]
The [specification](https://editorconfig-specification.readthedocs.io/#supported-pairs) says (emphasis added):
> `insert_final_newline`: Set to `true` ensure file ends with a newline when saving and `false` **to ensure it doesn’t.**

However:
- the Vim plugin, [back in 2014](https://github.com/editorconfig/editorconfig-vim/commit/a0250707b08c7671ac2391842e9b084a84f72343), treated `false` as "leave untouched", not "remove trailing newline".
- @treyhunner [said in 2012](https://github.com/editorconfig/editorconfig/issues/59#issuecomment-7508026) that "If `insert_final_newline = false` the text editor is simply instructed to *not* auto-insert final newlines".

Which is it?  Does `insert_final_newline == false` mean:
1. Don't touch the last EOL (as implemented); or
2. Strip the last EOL (as specified)

?

I think it's option 1, and that the specification should be updated, but I'm not sure.  Thanks!

Motivating example: https://github.com/editorconfig/editorconfig-vim/issues/136

Relates to #59 and #151

**Comment by xuhdev** [comment]
I remember that the Vim plugin implementation should be correct. However, editors have limitations and enforcing either should lead to some editors not being able to do so.

**Comment by ffes** [comment]
You are right, @cxw42. Clarifying this text a bit more in the specs could be useful. But as @xuhdev says, editors have their limitations.

FYI, the Notepad++ plugin deletes the final newline(s) when `insert_final_newline = false`.

**Comment by cxw42** [comment]
I agree not all editors can implement all properties.  However, I still think we should at least make sure the specification lists the _intended_ behaviour of the property, so that plugin authors know the goal.

**Comment by drenze-athene** [comment]
So, to validate: The **intended** behavior of `insert_final_newline = true` is:

* `true`: Ensure that the file ends with a newline.
* `false`: Strip the final newline from the file.

The third use case is: Multiple newlines at the end of the file (e.g., `\n\n\n`). If `true`, does it strip newlines and leave the file with one and only one newline at the end of the file (i.e., `\n`), or does it leave it untouched?

I requested this feature in #483, and was told that this issue indicates that this is the expected behavior, but I don't read it that way, so perhaps I was unclear in my feature request.

Can someone please provided clarification?

Thx.

**Comment by xuhdev** [comment]
I don't think they are the same. However, I don't think we'll implement exactly `N` new lines at EOF because this is not a feature that's widely available among editors (because, this feature doesn't have many use cases).

However, this issue should still be relevant to what you are looking for (exactly 1 EOL or only must end with EOL?)

**Comment by cxw42** [comment]
I think I would be surprised if a property called "**insert**_final_newline" caused newlines to be **removed** :) .  I believe we should update the specification to say that `insert_final_newline = false` means don't touch the newlines at EOF.  I suggest:
```diff
 insert_final_newline: Set to true
+to
  ensure file ends with a newline when 
-saving
+saving.
-and false to ensure it doesn’t.
+If it is set to false, the newline(s) at the end of the file will neither be checked nor modified.
```
What say you all?

**Comment by xuhdev** [comment]
Actually I don't think this is true. If the existing file has only 1 final newline, some editors will delete this final newline.

What if the file has multiple newlines and the plugin tells the editor to not add final new lines? I guess editors will do this in wildly different ways (not verified).

**Comment by xuhdev** [comment]
Yes, this is a bit confusing. Vim plugin doesn't do it partly because there's no way (or no clean way) to enforce no EOF of a file.

So what has really caused the confusion here may be that the spec should explicitly state that a plugin doesn't have to implement each single requirement -- they need to do their best in the context of the working editor. This is kind of obvious for developers in this circle but may not be obvious for anyone new to this system. What do you think?

**Comment by renzedj** [comment]
> So what has really caused the confusion here may be that the spec should explicitly state that a plugin doesn't have to implement each single requirement -- they need to do their best in the context of the working editor. This is kind of obvious for developers in this circle but may not be obvious for anyone new to this system. What do you think?

I would also add the requirement that where there is a difference between the specification and the behavior of the plugin, that difference MUST be documented. That would help reduce noise.

**Comment by gpanders** [comment]
To second what @justinmk said, I would like to propose that the EditorConfig specification be amended in one of two ways:

1. Clarify and enforce
   * The wording of the specification remains the same, but the current option (`insert_final_newline`) is deprecated and replaced with something like `enforce_final_newline=[insert|remove]` or similar
2. Relax
   * Modify the text in the specification to read "Set to true ensure file ends with a newline". No other text is needed: this is a boolean option, so if it's not set to true then the obvious implication is that it simply doesn't do anything and is as good as not setting the option at all

Taking the position that "editors should just implement the spec for whatever makes sense to them" diminishes the value of having a specification at all, and we will very soon (if we haven't already) run into "well it works in my editor" kind of issues. The whole point of having EditorConfig is to have common behavior enforced across editors. This ambiguity in the specification is detrimental to that objective.

**Comment by lewis6991** [comment]
> The third use case is: Multiple newlines at the end of the file (e.g., \n\n\n). If true, is the intent of the specification for insert_final_newline to strip newlines

Given this situation, I think it's pretty clear that `insert_final_newline==false` shouldn't mean "strip any trailing newlines from the file". The wording of the spec needs to be clarified and relaxed.

**Comment by cxw42** [comment]
I agree with
> 2. Relax
>    * Modify the text in the specification to read "Set to true ensure file ends with a newline"

**Comment by Technical-13** [comment]
GitHub editor does not strip any **EOF** `\n` when `insert_final_newline = false`.
Perhaps it should be changed that `insert_final_newline` only adds a newline if there's none and leaves untouched if there's at least one and a new `strip_final_newlines` should be created to strip all newlines if there's any and leave untouched if there's none?

**Comment by florianb** [comment]
Hi @Technical-13 - also regarding your issue #518 - to be clear, if we decide to write down how the EditorConfig should be interpreted, the GitHub-editor's maintainers would have to implement it. I fear there's nothing we can do to make them respect EditorConfig settings.

**Comment by Technical-13** [comment]
> Hi @Technical-13 - also regarding your issue #518 - to be clear, if we decide to write down how the EditorConfig should be interpreted, the GitHub-editor's maintainers would have to implement it. I fear there's nothing we can do to make them respect EditorConfig settings.

Obviously.  I feel the path of least resistance is to get a defined interpretation and then get them to implement it.
